File: src/services/ses.js
// services/ses.js — AWS SES emulator
import { store, randomId, arn } from '../store.js';
import { jsonResponse, errorJson, xmlResponse, escapeXml } from '../middleware/response.js';

function parseBody(req) {
  try { return JSON.parse(req.rawBody || '{}'); } catch { return {}; }
}
function parseForm(req) {
  return Object.fromEntries(new URLSearchParams(req.rawBody || '').entries());
}

const TARGET_MAP = {
  'AmazonSimpleEmailService.SendEmail':          sendEmailJson,
  'AmazonSimpleEmailService.SendRawEmail':       sendRawEmail,
  'AmazonSimpleEmailService.VerifyEmailIdentity': verifyEmail,
  'AmazonSimpleEmailService.VerifyEmailAddress':  verifyEmail,
  'AmazonSimpleEmailService.ListIdentities':      listIdentities,
  'AmazonSimpleEmailService.ListVerifiedEmailAddresses': listIdentities,
  'AmazonSimpleEmailService.DeleteIdentity':      deleteIdentity,
  'AmazonSimpleEmailService.GetSendQuota':        getSendQuota,
  'AmazonSimpleEmailService.GetSendStatistics':   getSendStats,
  'AmazonSimpleEmailService.GetIdentityVerificationAttributes': getVerificationAttrs,
};

export function handler(req, res) {
  const target = req.headers['x-amz-target'] || '';
  const fn = TARGET_MAP[target];
  if (fn) return fn(req, res);

  // Form-encoded (old SES style)
  const params = parseForm(req);
  const action = params.Action || new URLSearchParams(req.url.split('?')[1] || '').get('Action');
  if (action === 'SendEmail')           return sendEmailForm(req, res, params);
  if (action === 'VerifyEmailIdentity') return verifyEmailForm(req, res, params);
  if (action === 'ListIdentities')      return listIdentitiesForm(req, res);
  if (action === 'GetSendQuota')        return getSendQuotaForm(req, res);

  return errorJson(res, 400, 'InvalidAction', `Unknown SES action: ${target || action}`);
}

function mkEmail(from, to, subject, body, html) {
  const id = randomId(36);
  const email = {
    messageId: id,
    from,
    to:      Array.isArray(to) ? to : [to],
    subject: subject || '(no subject)',
    body:    body    || '',
    html:    html    || '',
    sent:    Date.now(),
  };
  store.ses.emails.unshift(email);
  if (store.ses.emails.length > 500) store.ses.emails.pop();
  store.ses.sent++;
  store.addTrail({ method: 'POST', path: '/ses/SendEmail', status: 200, latency: 8 });
  return id;
}

function sendEmailJson(req, res) {
  const b = parseBody(req);
  const from    = b.Source || b.FromEmailAddress || 'noreply@mockcloud.local';
  const to      = b.Destination?.ToAddresses || b.Destination?.to || [];
  const subject = b.Message?.Subject?.Data || b.Subject || '';
  const body    = b.Message?.Body?.Text?.Data || '';
  const html    = b.Message?.Body?.Html?.Data || '';
  const id      = mkEmail(from, to, subject, body, html);
  jsonResponse(res, 200, { MessageId: id });
}

function sendRawEmail(req, res) {
  const b  = parseBody(req);
  const id = mkEmail('raw@mockcloud.local', [], '(raw email)', b.RawMessage?.Data || '', '');
  jsonResponse(res, 200, { MessageId: id });
}

function verifyEmail(req, res) {
  const b     = parseBody(req);
  const email = b.EmailAddress || b.Identity;
  if (email) store.ses.identities[email] = { email, status: 'Success', verified: true };
  jsonResponse(res, 200, {});
}

function deleteIdentity(req, res) {
  const b = parseBody(req);
  delete store.ses.identities[b.Identity];
  jsonResponse(res, 200, {});
}

function listIdentities(req, res) {
  jsonResponse(res, 200, { Identities: Object.keys(store.ses.identities) });
}

function getSendQuota(req, res) {
  jsonResponse(res, 200, { Max24HourSend: 50000, MaxSendRate: 14, SentLast24Hours: store.ses.sent });
}

function getSendStats(req, res) {
  jsonResponse(res, 200, { SendDataPoints: [] });
}

function getVerificationAttrs(req, res) {
  const b = parseBody(req);
  const attrs = {};
  for (const id of (b.Identities || [])) {
    attrs[id] = { VerificationStatus: store.ses.identities[id] ? 'Success' : 'Pending' };
  }
  jsonResponse(res, 200, { VerificationAttributes: attrs });
}

// Form-encoded variants
function sendEmailForm(req, res, p) {
  const id = mkEmail(p.Source, p['Destination.ToAddresses.member.1'], p['Message.Subject.Data'], p['Message.Body.Text.Data'], p['Message.Body.Html.Data']);
  xmlResponse(res, 200, `<?xml version="1.0"?><SendEmailResponse><SendEmailResult><MessageId>${id}</MessageId></SendEmailResult></SendEmailResponse>`);
}
function verifyEmailForm(req, res, p) {
  if (p.EmailAddress) store.ses.identities[p.EmailAddress] = { email: p.EmailAddress, status: 'Success', verified: true };
  xmlResponse(res, 200, `<?xml version="1.0"?><VerifyEmailIdentityResponse><VerifyEmailIdentityResult/></VerifyEmailIdentityResponse>`);
}
function listIdentitiesForm(req, res) {
  const members = Object.keys(store.ses.identities).map((e, i) => `<member>${escapeXml(e)}</member>`).join('');
  xmlResponse(res, 200, `<?xml version="1.0"?><ListIdentitiesResponse><ListIdentitiesResult><Identities>${members}</Identities></ListIdentitiesResult></ListIdentitiesResponse>`);
}
function getSendQuotaForm(req, res) {
  xmlResponse(res, 200, `<?xml version="1.0"?><GetSendQuotaResponse><GetSendQuotaResult><Max24HourSend>50000</Max24HourSend><MaxSendRate>14</MaxSendRate><SentLast24Hours>${store.ses.sent}</SentLast24Hours></GetSendQuotaResult></GetSendQuotaResponse>`);
}

// ── Inbound receipt rules (control-plane driven) ────────────────────────────
// A local mock can't receive real SMTP, so inbound mail is simulated via the
// control plane (POST /mockcloud/ses/inbound). For each enabled receipt rule
// whose recipients match, run its actions — reusing the S3 / SNS / Lambda
// delivery paths so an inbound email can land an object, fan out a
// notification, or invoke a function, exactly like real SES receipt rules.
//   rule:   { name, enabled?, recipients: [addr|domain], actions: [Action] }
//   Action: { type:'s3', bucket, objectKeyPrefix? }
//         | { type:'sns', topicArn }
//         | { type:'lambda', functionArn }
export async function deliverInboundEmail({ from, to, subject, body } = {}) {
  const recipients = (Array.isArray(to) ? to : [to]).filter(Boolean);
  const messageId  = randomId(36);
  const rawEmail   = `From: ${from || ''}\r\nTo: ${recipients.join(', ')}\r\nSubject: ${subject || ''}\r\n\r\n${body || ''}`;

  const matched = [];
  for (const rule of store.ses.receiptRules || []) {
    if (rule.enabled === false) continue;
    if (!recipientMatches(rule.recipients, recipients)) continue;
    matched.push(rule.name);
    for (const action of rule.actions || []) {
      try { await runReceiptAction(action, { messageId, from, recipients, subject, rawEmail }); }
      catch (e) { console.warn(`[SES] receipt action ${action?.type} failed:`, e.message); }
    }
  }
  // Record the inbound message in the same log the UI lists outbound from.
  store.ses.emails.unshift({ messageId, direction: 'inbound', from, to: recipients, subject: subject || '(no subject)', body: body || '', matchedRules: matched, sent: Date.now() });
  if (store.ses.emails.length > 500) store.ses.emails.pop();
  return { messageId, matched };
}

// Empty/absent recipient list on a rule matches everything. Otherwise match a
// full address, a domain (`example.com`), or an address ending in `@domain`.
function recipientMatches(ruleRecipients, recipients) {
  if (!ruleRecipients || ruleRecipients.length === 0) return true;
  return recipients.some(addr =>
    ruleRecipients.some(r => addr === r || addr.endsWith('@' + r) || addr.endsWith(r)));
}

async function runReceiptAction(action, ctx) {
  switch ((action.type || '').toLowerCase()) {
    case 's3': {
      const { putObjectToBucket } = await import('./s3.js');
      const key = `${action.objectKeyPrefix || ''}${ctx.messageId}`;
      putObjectToBucket(action.bucket, key, Buffer.from(ctx.rawEmail), 'message/rfc822');
      return;
    }
    case 'sns': {
      const topic = store.sns.topics[action.topicArn];
      if (!topic) return;
      const { fanoutSnsMessage } = await import('./sns.js');
      topic.published = (topic.published || 0) + 1;
      const notification = JSON.stringify({
        notificationType: 'Received',
        mail: { messageId: ctx.messageId, source: ctx.from, destination: ctx.recipients, commonHeaders: { subject: ctx.subject } },
      });
      await fanoutSnsMessage(topic, { msgId: randomId(36), message: notification, subject: 'Amazon SES Email Receipt Notification' });
      return;
    }
    case 'lambda': {
      const { invokeLambda } = await import('./lambda.js');
      const fnName = action.functionArn.split(':').pop();
      const event = { Records: [{
        eventSource: 'aws:ses', eventVersion: '1.0',
        ses: {
          mail: { messageId: ctx.messageId, source: ctx.from, destination: ctx.recipients, commonHeaders: { subject: ctx.subject, from: [ctx.from], to: ctx.recipients } },
          receipt: { recipients: ctx.recipients, action: { type: 'Lambda', functionArn: action.functionArn } },
        },
      }] };
      invokeLambda(fnName, event, { source: 'ses' }).catch(() => {});
      return;
    }
  }
}

File: internal/services/ses/ses.go
// Package ses — port of src/services/ses.js: outbound SendEmail/SendRawEmail
// over both the JSON-target and query/XML protocols (SES answers XML on the
// form path — the response strings below are byte-for-byte Node's), identity
// verification, quotas/statistics, and control-plane-driven inbound receipt
// rules (inbound.go) that fan out to S3 / SNS / Lambda.
package ses

import (
	"net/http"
	"net/url"
	"sort"
	"strconv"
	"strings"

	"github.com/mockcloud/mockcloud/internal/config"
	"github.com/mockcloud/mockcloud/internal/httpapi"
	"github.com/mockcloud/mockcloud/internal/jsnum"
	"github.com/mockcloud/mockcloud/internal/protocol/respond"
	"github.com/mockcloud/mockcloud/internal/state"
	"github.com/mockcloud/mockcloud/internal/store"
)

type Service struct {
	st  *store.Store
	cfg *config.Config
	// Seams wired by dispatch.New — both called OUTSIDE the store lock.
	// FanoutSNS is synchronous (Node awaited fanoutSnsMessage for receipt
	// actions); InvokeLambda is fire-and-forget (source 'ses').
	FanoutSNS    func(topicArn, msgID, message, subject string)
	InvokeLambda func(fnName, eventJSON string)
}

func New(st *store.Store, cfg *config.Config) *Service { return &Service{st: st, cfg: cfg} }

func (svc *Service) Handler(w http.ResponseWriter, r *httpapi.Request) {
	target := r.Header.Get("x-amz-target")
	if op, ok := strings.CutPrefix(target, "AmazonSimpleEmailService."); ok {
		b := r.JSONBody()
		switch op {
		case "SendEmail":
			svc.sendEmailJSON(w, b)
			return
		case "SendRawEmail":
			svc.sendRawEmailJSON(w, b)
			return
		case "VerifyEmailIdentity", "VerifyEmailAddress":
			svc.verifyEmailJSON(w, b)
			return
		case "ListIdentities", "ListVerifiedEmailAddresses":
			svc.listIdentitiesJSON(w)
			return
		case "DeleteIdentity":
			svc.deleteIdentityJSON(w, b)
			return
		case "GetSendQuota":
			svc.getSendQuotaJSON(w)
			return
		case "GetSendStatistics":
			respond.JSON(w, 200, map[string]any{"SendDataPoints": []any{}})
			return
		case "GetIdentityVerificationAttributes":
			svc.getVerificationAttrsJSON(w, b)
			return
		}
	}

	// Form-encoded (old SES style)
	params, _ := url.ParseQuery(string(r.RawBody))
	action := params.Get("Action")
	if action == "" {
		action = r.Query.Get("Action")
	}
	switch action {
	case "SendEmail":
		svc.sendEmailForm(w, params)
		return
	case "SendRawEmail":
		svc.sendRawEmailForm(w, params)
		return
	case "VerifyEmailIdentity", "VerifyEmailAddress":
		svc.verifyEmailForm(w, params, action)
		return
	case "ListIdentities":
		svc.listIdentitiesForm(w, "ListIdentities", "Identities")
		return
	case "ListVerifiedEmailAddresses":
		svc.listIdentitiesForm(w, "ListVerifiedEmailAddresses", "VerifiedEmailAddresses")
		return
	case "DeleteIdentity":
		svc.deleteIdentityForm(w, params)
		return
	case "GetSendQuota":
		svc.getSendQuotaForm(w)
		return
	case "GetSendStatistics":
		respond.XML(w, 200, `<?xml version="1.0"?><GetSendStatisticsResponse><GetSendStatisticsResult><SendDataPoints/></GetSendStatisticsResult></GetSendStatisticsResponse>`)
		return
	case "GetIdentityVerificationAttributes":
		svc.getVerificationAttrsForm(w, params)
		return
	}

	// Query-protocol callers expect an XML <ErrorResponse>; JSON targets a
	// {__type}. (Node: `${target || action}` — both empty renders "null".)
	if action != "" && target == "" {
		respond.ErrorXML(w, 400, "InvalidAction", "Unknown SES action: "+action)
		return
	}
	label := target
	if label == "" {
		label = action
	}
	if label == "" {
		label = "null"
	}
	respond.ErrorJSON(w, 400, "InvalidAction", "Unknown SES action: "+label)
}

// mkEmail records an outbound email (ring capped at 500) and bumps the sent
// counter. `to` non-arrays are wrapped, like Node's mkEmail.
func (svc *Service) mkEmail(from, to, subject, body, html any) string {
	id := state.RandomID(36)
	toArr, ok := to.([]any)
	if !ok {
		toArr = []any{to}
	}
	email := map[string]any{
		"messageId": id,
		"subject":   orDefault(subject, "(no subject)"),
		"body":      orDefault(body, ""),
		"html":      orDefault(html, ""),
		"sent":      state.NowMs(),
	}
	email["to"] = toArr
	setDefined(email, "from", from)
	svc.st.With(func(s *state.State) {
		s.SES.Emails = append([]map[string]any{email}, s.SES.Emails...)
		if len(s.SES.Emails) > 500 {
			s.SES.Emails = s.SES.Emails[:500]
		}
		s.SES.Sent++
		s.AddTrail(map[string]any{"method": "POST", "path": "/ses/SendEmail", "status": 200, "latency": 8})
	})
	return id
}

// orDefault — JS `v || def`.
func orDefault(v, def any) any {
	if jsnum.Truthy(v) {
		return v
	}
	return def
}

// setDefined mirrors JSON.stringify dropping undefined-valued keys.
func setDefined(m map[string]any, key string, v any) {
	if v != nil {
		m[key] = v
	}
}

// dig — optional chaining over the any-tree (a?.b?.c).
func dig(m map[string]any, keys ...string) any {
	var cur any = m
	for _, k := range keys {
		mm, ok := cur.(map[string]any)
		if !ok {
			return nil
		}
		cur = mm[k]
	}
	return cur
}

// ── JSON-target handlers ────────────────────────────────────────────────────

func (svc *Service) sendEmailJSON(w http.ResponseWriter, b map[string]any) {
	from := orDefault(orDefault(b["Source"], b["FromEmailAddress"]), "noreply@mockcloud.local")
	to := orDefault(orDefault(dig(b, "Destination", "ToAddresses"), dig(b, "Destination", "to")), []any{})
	subject := orDefault(orDefault(dig(b, "Message", "Subject", "Data"), b["Subject"]), "")
	body := orDefault(dig(b, "Message", "Body", "Text", "Data"), "")
	html := orDefault(dig(b, "Message", "Body", "Html", "Data"), "")
	id := svc.mkEmail(from, to, subject, body, html)
	respond.JSON(w, 200, map[string]any{"MessageId": id})
}

func (svc *Service) sendRawEmailJSON(w http.ResponseWriter, b map[string]any) {
	id := svc.mkEmail("raw@mockcloud.local", []any{}, "(raw email)", orDefault(dig(b, "RawMessage", "Data"), ""), "")
	respond.JSON(w, 200, map[string]any{"MessageId": id})
}

func (svc *Service) verifyEmailJSON(w http.ResponseWriter, b map[string]any) {
	email := orDefault(b["EmailAddress"], b["Identity"])
	if jsnum.Truthy(email) {
		svc.st.With(func(s *state.State) {
			s.SES.Identities[jsnum.ToString(email)] = map[string]any{"email": email, "status": "Success", "verified": true}
		})
	}
	respond.JSON(w, 200, map[string]any{})
}

func (svc *Service) deleteIdentityJSON(w http.ResponseWriter, b map[string]any) {
	svc.st.With(func(s *state.State) { delete(s.SES.Identities, httpapi.Str(b, "Identity")) })
	respond.JSON(w, 200, map[string]any{})
}

func (svc *Service) listIdentitiesJSON(w http.ResponseWriter) {
	ids := []string{}
	svc.st.With(func(s *state.State) {
		for k := range s.SES.Identities {
			ids = append(ids, k)
		}
	})
	sort.Strings(ids)
	respond.JSON(w, 200, map[string]any{"Identities": ids})
}

func (svc *Service) getSendQuotaJSON(w http.ResponseWriter) {
	var sent float64
	svc.st.With(func(s *state.State) { sent = s.SES.Sent })
	respond.JSON(w, 200, map[string]any{"Max24HourSend": 50000, "MaxSendRate": 14, "SentLast24Hours": sent})
}

func (svc *Service) getVerificationAttrsJSON(w http.ResponseWriter, b map[string]any) {
	attrs := map[string]any{}
	ids, _ := b["Identities"].([]any)
	svc.st.With(func(s *state.State) {
		for _, raw := range ids {
			id := jsnum.ToString(raw)
			status := "Pending"
			if _, ok := s.SES.Identities[id]; ok {
				status = "Success"
			}
			attrs[id] = map[string]any{"VerificationStatus": status}
		}
	})
	respond.JSON(w, 200, map[string]any{"VerificationAttributes": attrs})
}

// ── Form-encoded variants (XML responses — byte-for-byte Node's strings) ────

// formVal — a form field as an any-tree value: present → string, absent →
// nil (Node's undefined, so mkEmail's `|| fallback`s behave identically).
func formVal(params url.Values, key string) any {
	if params.Has(key) {
		return params.Get(key)
	}
	return nil
}

func (svc *Service) sendEmailForm(w http.ResponseWriter, p url.Values) {
	id := svc.mkEmail(formVal(p, "Source"), formVal(p, "Destination.ToAddresses.member.1"),
		formVal(p, "Message.Subject.Data"), formVal(p, "Message.Body.Text.Data"), formVal(p, "Message.Body.Html.Data"))
	respond.XML(w, 200, `<?xml version="1.0"?><SendEmailResponse><SendEmailResult><MessageId>`+id+`</MessageId></SendEmailResult></SendEmailResponse>`)
}

func (svc *Service) sendRawEmailForm(w http.ResponseWriter, p url.Values) {
	id := svc.mkEmail("raw@mockcloud.local", []any{}, "(raw email)", orDefault(formVal(p, "RawMessage.Data"), ""), "")
	respond.XML(w, 200, `<?xml version="1.0"?><SendRawEmailResponse><SendRawEmailResult><MessageId>`+id+`</MessageId></SendRawEmailResult></SendRawEmailResponse>`)
}

// VerifyEmailIdentity and the legacy VerifyEmailAddress differ only in the
// response wrapper, so both route here with their action name.
func (svc *Service) verifyEmailForm(w http.ResponseWriter, p url.Values, name string) {
	if email := p.Get("EmailAddress"); email != "" {
		svc.st.With(func(s *state.State) {
			s.SES.Identities[email] = map[string]any{"email": email, "status": "Success", "verified": true}
		})
	}
	respond.XML(w, 200, `<?xml version="1.0"?><`+name+`Response><`+name+`Result/></`+name+`Response>`)
}

// Same for ListIdentities / ListVerifiedEmailAddresses (different list element).
func (svc *Service) listIdentitiesForm(w http.ResponseWriter, name, listTag string) {
	ids := []string{}
	svc.st.With(func(s *state.State) {
		for k := range s.SES.Identities {
			ids = append(ids, k)
		}
	})
	sort.Strings(ids)
	var members strings.Builder
	for _, e := range ids {
		members.WriteString("<member>" + respond.EscapeXML(e) + "</member>")
	}
	respond.XML(w, 200, `<?xml version="1.0"?><`+name+`Response><`+name+`Result><`+listTag+`>`+members.String()+`</`+listTag+`></`+name+`Result></`+name+`Response>`)
}

func (svc *Service) deleteIdentityForm(w http.ResponseWriter, p url.Values) {
	svc.st.With(func(s *state.State) { delete(s.SES.Identities, p.Get("Identity")) })
	respond.XML(w, 200, `<?xml version="1.0"?><DeleteIdentityResponse><DeleteIdentityResult/></DeleteIdentityResponse>`)
}

func (svc *Service) getSendQuotaForm(w http.ResponseWriter) {
	var sent float64
	svc.st.With(func(s *state.State) { sent = s.SES.Sent })
	respond.XML(w, 200, `<?xml version="1.0"?><GetSendQuotaResponse><GetSendQuotaResult><Max24HourSend>50000</Max24HourSend><MaxSendRate>14</MaxSendRate><SentLast24Hours>`+jsnum.Format(sent)+`</SentLast24Hours></GetSendQuotaResult></GetSendQuotaResponse>`)
}

func (svc *Service) getVerificationAttrsForm(w http.ResponseWriter, p url.Values) {
	var entries strings.Builder
	svc.st.With(func(s *state.State) {
		for i := 1; p.Has("Identities.member." + strconv.Itoa(i)); i++ {
			id := p.Get("Identities.member." + strconv.Itoa(i))
			status := "Pending"
			if _, ok := s.SES.Identities[id]; ok {
				status = "Success"
			}
			entries.WriteString("<entry><key>" + respond.EscapeXML(id) + "</key><value><VerificationStatus>" + status + "</VerificationStatus></value></entry>")
		}
	})
	respond.XML(w, 200, `<?xml version="1.0"?><GetIdentityVerificationAttributesResponse><GetIdentityVerificationAttributesResult><VerificationAttributes>`+entries.String()+`</VerificationAttributes></GetIdentityVerificationAttributesResult></GetIdentityVerificationAttributesResponse>`)
}
